### PR TITLE
feat!: genesis

### PR DESCRIPTION
**Pull request by Gozala**
Mostly it was started as adaptation of the https://github.com/rvagg/js-fil-utils into ESM, however testing against go implementation uncovered [various incompatibilities](https://filecoinproject.slack.com/archives/C02BZPRS9HP/p1684862906021219) which lead me to adapting to  the approach used by https://github.com/filecoin-project/go-data-segment

Test data sets have been generated by feeding deterministically random data into [stream-commp](https://github.com/filecoin-project/go-fil-commp-hashhash/tree/master/cmd/stream-commp#usage-example) binary and capturing outputs into .csv file.

### CODE REVIEW

File: script/make-fixtures.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { vector } from './commp.vector.js'
```
Comment (vasco-santos) [inline script/make-fixtures.js]: Looks like this file does not exist?

File: src/api.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+ * Implementers of the `Read` interface are called "readers". Readers
+ * allow for reading bytes from an underlying source.
+ *
+ * Readers are defined by one required method, `read`. Each call to `read`
+ * will attempt to pull bytes from this source into a provided buffer. A
+ * number of high level functions are implemented in terms of `read`, giving
+ * users a number of ways to read bytes while only needing to implement a
+ * single method.
+ *
+ * This based on [std::io::Read] trait in Rust.
```
Comment (vasco-santos) [inline src/api.ts]:
```suggestion
 * This is based on [std::io::Read] trait in Rust.
```

File: src/api.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+ */
+interface InferenceError<message> {
+  [Marker]: never & message
+}
+
+/**
+ * Defines result type as per invocation spec
+ *
+ * @see https://github.com/ucan-wg/invocation/#6-result
+ */
```
Comment (vasco-santos) [inline src/api.ts]: Might not make sense to call out ucan invocation spec here?

File: src/buffer.js [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+      // because it preceeds the end & we had not found start yet.
+      // In such case we create a view with only single segment of bytes
+      // in the range.
+      if (end <= nextOffset) {
+        const range = segment.subarray(start - offset, end - offset)
+        segments.push(range)
+        byteLength = range.byteLength
+        break
+      }
+      // If start offeset falls with in current range (but not the end)
```
Comment (vasco-santos) [inline src/buffer.js]:
```suggestion
      // If start offset falls with in current range (but not the end)
```

File: src/commp.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  return new CommP({ tree, size: source.byteLength })
+}
+
+/**
+ * @see https://github.com/filecoin-project/go-fil-commcid/blob/d41df56b4f6a934316028e4d4b93fb220674801d/commcid.go#L141-L144
+ * @see https://github.com/filecoin-project/go-fil-commcid/blob/d41df56b4f6a934316028e4d4b93fb220674801d/commcid.go#L79-L81
+ *
+ * @param {Uint8Array} root
+ * @param {object} options
+ * @param {import('multiformats/codecs/interface').BlockCodec<any, any>} options.codec
```
Comment (vasco-santos) [inline src/commp.js]:
```suggestion
```

File: src/proof.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
+import * as API from './api.js'
+import { sha256 } from 'multiformats/hashes/sha2'
+
+/**
+ * @typedef {API.ProofData} ProofData
+ */
+
```
Comment (vasco-santos) [inline src/proof.js]:
```suggestion
```

do we need it?

File: src/commp.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * @see https://github.com/ipfs/go-cid/blob/829c826f6be23320846f4b7318aee4d17bf8e094/cid.go#L104
+ */
+const FilCommitmentUnsealed = 0xf101
+
+/**
+ * @param {Uint8Array} source
+ */
+export const build = async (source) => {
+  if (source.byteLength <= 64) {
```
Comment (vasco-santos) [inline src/commp.js]: Should we make 64 a constant?

File: src/proof.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+/**
+ * @param {API.MerkleTreeNode} subtree
+ * @param {API.ProofData} proofData
+ * @returns {Promise<API.Result<API.MerkleTreeNode, Error>>}
+ */
+export async function computeRoot(subtree, proofData) {
+  if (subtree === null) {
+    return { error: new Error('nil subtree cannot be used') }
+  }
+  if (depth(proofData) > 63) {
```
Comment (vasco-santos) [inline src/proof.js]: Should we have a constant here for 63?

File: src/proof.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+  }
+  return { ok: undefined }
+}
+
+/**
+ * @param {API.ProofData} proofData
+ * @returns {API.Result<void, Error>}
+ */
+export function validateProofStructure(proofData) {
+  // In the Go implementation, this function does not perform any checks and always returns nil error
```
Comment (vasco-santos) [inline src/proof.js]: interesting. Is this some upcoming work planned? Or this might not be needed?

File: src/proof.js [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  return truncatedHash(payload)
+}
+
+/**
+ *
+ * @param {API.MerkleTreeNode} node
+ * @returns {API.MerkleTreeNode}
+ */
+export function truncate(node) {
+  node[32 - 1] &= 0b00111111
```
Comment (vasco-santos) [inline src/proof.js]: Should we use a constant for `0b00111111` to improve readability?

File: script/make-fixtures.js (hunk 1 above)
Comment (vasco-santos) [reply script/make-fixtures.js]: Ok it runs in prepare script. Could be cool to add more info on where it comes from

**Review by vasco-santos — COMMENTED** [review]
Overall looking good! Great work 🏅 

Left a few suggestions/details inline. In addition, could we get README updated and some documentation around the scripts we have here

File: src/api.ts [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+ * where some error message would be useful.
+ */
+interface InferenceError<message> {
+  [Marker]: never & message
+}
+
+/**
+ * Defines result type as per invocation spec
+ *
+ * @see https://github.com/ucan-wg/invocation/#6-result
```
Comment (Gozala) [inline src/api.ts]:
```suggestion
 * Defines result type in an idiomatic IPLD representation.
```

**Pull request merged**